sw/device/silicon_creator/lib/sigverify/sphincsplus/hash_shake.c

```diff
@@ -0,0 +1,88 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Derived from code in the SPHINCS+ reference implementation (CC0 license):
+// https://github.com/sphincs/sphincsplus/blob/ed15dd78658f63288c7492c00260d86154b84637/ref/hash_shake.c
+
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/silicon_creator/lib/drivers/kmac.h"
+#include "sw/device/silicon_creator/lib/sigverify/sphincsplus/address.h"
+#include "sw/device/silicon_creator/lib/sigverify/sphincsplus/hash.h"
+#include "sw/device/silicon_creator/lib/sigverify/sphincsplus/params.h"
+#include "sw/device/silicon_creator/lib/sigverify/sphincsplus/utils.h"
+
+enum {
+  /**
+   * Number of bits needed to represent the `tree` field.
+   */
+  kSpxTreeBits = kSpxTreeHeight * (kSpxD - 1),
+  /**
+   * Number of bytes needed to represent the `tree` field.
+   */
+  kSpxTreeBytes = (kSpxTreeBits + 7) / 8,
+  /**
+   * Number of bits needed to represent a leaf index.
+   */
+  kSpxLeafBits = kSpxTreeHeight,
+  /**
+   * Number of bytes needed to represent a leaf index.
+   */
+  kSpxLeafBytes = (kSpxLeafBits + 7) / 8,
+  /**
+   * Number of bytes needed for the message digest.
+   */
+  kSpxDigestBytes = kSpxForsMsgBytes + kSpxTreeBytes + kSpxLeafBytes,
+  /**
+   * Number of 32-bit words needed for the message digest.
+   *
+   * Rounded up if necessary.
+   */
+  kSpxDigestWords = (kSpxDigestBytes + sizeof(uint32_t) - 1) / sizeof(uint32_t),
+};
+
+static_assert(
+    kSpxTreeBits <= 64,
+    "For given height and depth, 64 bits cannot represent all subtrees.");
```

Comment:
Should we add a static assert for `kSpxLeafBits` as well given that there is another buffer to `uint64_t` conversion being used in the code?